### PR TITLE
[json-rpc] generic endpoint for retrieving all resources

### PR DESCRIPTION
**Pull request by sblackshear**
in JSON format (or any other format serde can produce from an account.

This PR adds a new endpoint for displaying account contents in a human-readable fashion. The endpoint leverages the existing `MoveResourceViewer` to automatically discover the structure of each resource using the on-chain modules and render the resource. An example probably explains best:

```
curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"get_resources","params":["0000000000000000000000000a550c18"],"id":1}' http://0.0.0.0:8080 | jq .

<snip>
  "0x1::Event::EventHandleGenerator": {
      "counter": 18,
      "addr": "0000000000000000000000000a550c18"
    },
    "0x1::Roles::RoleId": {
      "role_id": 0
    },
    "0x1::SlidingNonce::SlidingNonce": {
      "min_nonce": 0,
      "nonce_mask": 0
    },
    "0x1::XDX::Reserve": {
      "mint_cap": {
        "dummy_field": false
      },
      "burn_cap": {
        "dummy_field": false
      },
      "preburn_cap": {
        "to_burn": {
          "value": 0
        }
      }
    }
```

Thanks to @bmwill for helping me untangle some nasty trait-related issues to get this code compiling.

### CODE REVIEW

**Comment by libra-action** [comment]
**This PR may have modified JSON-RPC server code.**

Breaking changes policy:

1. Changing JSON-RPC method signatures or removing/modifying fields
   in the response in /V1 will not be accepted.
2. Adding fields is ok.
3. Adding a new JSON-RPC method is ok.

Please ensure you have also done the following to update the docs:

1. Update json-rpc/API-CHANGELOG.md.
2. Add/update type/method documentation under json-rpc/docs.

**Comment by libra-action** [comment]
**LSR dependency change summary**:
```
target packages:
  A borrow-graph 0.0.1 (workspace, path 'language/borrow-graph')
    * features: [none]
  A bytecode-verifier 0.1.0 (workspace, path 'language/bytecode-verifier')
    * features: default
  A move-binary-format 0.1.0 (workspace, path 'language/move-binary-format')
    * features: default
  A move-vm-runtime 0.1.0 (workspace, path 'language/move-vm/runtime')
    * features: default
  A move-vm-types 0.1.0 (workspace, path 'language/move-vm/types')
    * features: default
  A fail 0.4.0 (direct third-party, crates.io)
    * features: [none]
  A petgraph 0.5.1 (direct third-party, crates.io)
    * features: default, graphmap, matrix_graph, stable_graph
  A sha3 0.9.1 (direct third-party, crates.io)
    * features: default, std
  A block-padding 0.2.1 (transitive third-party, crates.io)
    * features: [none]
  A fixedbitset 0.2.0 (transitive third-party, crates.io)
    * features: [none]
  A getrandom 0.1.16 (transitive third-party, crates.io) (other versions: 0.2.2)
    * features: std
  A hashbrown 0.9.1 (transitive third-party, crates.io)
    * features: raw
  A indexmap 1.6.2 (transitive third-party, crates.io)
    * features: [none]
  A keccak 0.1.0 (transitive third-party, crates.io)
    * features: [none]
  A rand 0.7.3 (transitive third-party, crates.io) (other versions: 0.8.3)
    * features: alloc, default, getrandom, getrandom_package, libc, std
  A rand_chacha 0.2.2 (transitive third-party, crates.io) (other versions: 0.3.0)
    * features: std
  A rand_core 0.5.1 (transitive third-party, crates.io) (other versions: 0.6.2)
    * features: alloc, getrandom, std
  M smallvec 1.6.1 (direct third-party, crates.io)
    * status changed from transitive third-party
    * (unchanged features: [none])
  M block-buffer 0.9.0 (transitive third-party, crates.io)
    * added features: block-padding
    * (unchanged features: [none])

host packages:
  A variant_count 1.1.0 (direct third-party, crates.io)
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A resource-viewer 0.1.0 (workspace, path 'language/tools/resource-viewer')
    * features: [none]
  M serde_json 1.0.64 (direct third-party, crates.io)
    * added features: arbitrary_precision
    * (unchanged features: default, std)
```

File: language/tools/resource-viewer/src/lib.rs [hunk 1]
```diff
@@ -226,6 +225,41 @@ fn pretty_print_ability_modifiers(f: &mut Formatter, abilities: AbilitySet) -> s
     Ok(())
 }
 
+impl serde::Serialize for AnnotatedMoveStruct {
+    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        // TODO: can we avoid Box::leak here?
```
Comment (sblackshear) [inline language/tools/resource-viewer/src/lib.rs:230]: @tnowacki @modocache: The issue I'm running into here feels similar to what we hit/fixed in the compiler: serde wants a `static` string for the type and field names, but I can't produce one because I'm reading them from on-chain. My naive attempt at a fix was going to be something like

```
struct AnnotatedMoveStructWithStringCache {
  s: AnnotatedMovedStruct
  cache: Vector<Box<String>>
```

+ using `Box::from_raw` to populate `cache` instead of using `leak`. But wanted to ask for your advice

Comment (modocache) [reply language/tools/resource-viewer/src/lib.rs:230]: I pulled down your code and confirmed that the new string interning package can achieve what you're asking for here, with a few caveats:

1. We discussed this in #8782, but currently `move_symbol_pool::Symbol::as_str` returns a `&str` with the same lifetime as the symbol, instead of a `&'static str`. The underlying `str` does indeed live for the entire duration of the program, though, so it could (and probably should) be marked static. Once that function signature is changed to return `&'static str`, it works for your use case here, by writing `Symbol::from(format!("{}", self.type_)).as_str()`. This would add the struct tag to the global string table and keep it around for the rest of the resource viewer's execution.
2. I don't truly understand why serde requires a static string reference in this method signature. https://www.joshmcguigan.com/blog/understanding-serde/ explains that the name of the struct being serialized is always known at compile-time. Does that mean that you're using `serialize_struct` here in a way that isn't intended?

Comment (sblackshear) [reply language/tools/resource-viewer/src/lib.rs:230]: Thanks for the detailed explanation!

To explain a bit more, I my use-case here is slightly different than the compiler in that:
- I would like for the strings I create to be deallocated at the same time as the `AnnotatedMoveStruct` rather than living for the duration of the program. This is important because this code will be called from `json-rpc`, which is a long-running server.
- I'm ok with allocating the same string multiple times (don't expect a lot of duplication), though avoiding this would be a nice bonus.

> don't truly understand why serde requires a static string reference in this method signature.

I think that this is because serde's code generation for the `Serialize` traits relies on compile-time macros that need to look at field/type names, but I'm not sure.

Comment (bmwill) [reply language/tools/resource-viewer/src/lib.rs:230]: There's one way I can think of to get around this, although its going to require two different code paths based on [`is_human_readable`](https://docs.rs/serde/1.0.127/serde/trait.Serializer.html#method.is_human_readable).

When `is_human_readable == true` this is generally going to be the "json" case, and here we don't actually care about serializing a named struct but instead we can serialize a [map](https://docs.rs/serde/1.0.127/serde/trait.Serializer.html#tymethod.serialize_map) using the non-static field names as keys.

When `is_human_readable == false` this is generally going to be the "bcs" case, and here we don't care about struct or field names since they aren't serialized in bcs, instead we can just serialize using a [tuple](https://docs.rs/serde/1.0.127/serde/trait.Serializer.html#tymethod.serialize_tuple).

Comment (sblackshear) [reply language/tools/resource-viewer/src/lib.rs:230]: Good idea, especially re: map in the JSON case! This type doesn't ever appear on-chain, so I'm not sure we need to do anything special for BCS, but I will do this anyway just in case.

**Comment by libra-action** [comment]
**LSR dependency change summary**:
```
target packages:
  A borrow-graph 0.0.1 (workspace, path 'language/borrow-graph')
    * features: [none]
  A bytecode-verifier 0.1.0 (workspace, path 'language/bytecode-verifier')
    * features: default
  A move-binary-format 0.1.0 (workspace, path 'language/move-binary-format')
    * features: default
  A move-vm-runtime 0.1.0 (workspace, path 'language/move-vm/runtime')
    * features: default
  A move-vm-types 0.1.0 (workspace, path 'language/move-vm/types')
    * features: default
  A fail 0.4.0 (direct third-party, crates.io)
    * features: [none]
  A petgraph 0.5.1 (direct third-party, crates.io)
    * features: default, graphmap, matrix_graph, stable_graph
  A sha3 0.9.1 (direct third-party, crates.io)
    * features: default, std
  A block-padding 0.2.1 (transitive third-party, crates.io)
    * features: [none]
  A fixedbitset 0.2.0 (transitive third-party, crates.io)
    * features: [none]
  A getrandom 0.1.16 (transitive third-party, crates.io) (other versions: 0.2.2)
    * features: std
  A hashbrown 0.9.1 (transitive third-party, crates.io)
    * features: raw
  A indexmap 1.6.2 (transitive third-party, crates.io)
    * features: [none]
  A keccak 0.1.0 (transitive third-party, crates.io)
    * features: [none]
  A rand 0.7.3 (transitive third-party, crates.io) (other versions: 0.8.3)
    * features: alloc, default, getrandom, getrandom_package, libc, std
  A rand_chacha 0.2.2 (transitive third-party, crates.io) (other versions: 0.3.0)
    * features: std
  A rand_core 0.5.1 (transitive third-party, crates.io) (other versions: 0.6.2)
    * features: alloc, getrandom, std
  M smallvec 1.6.1 (direct third-party, crates.io)
    * status changed from transitive third-party
    * (unchanged features: [none])
  M block-buffer 0.9.0 (transitive third-party, crates.io)
    * added features: block-padding
    * (unchanged features: [none])

host packages:
  A variant_count 1.1.0 (direct third-party, crates.io)
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A resource-viewer 0.1.0 (workspace, path 'language/tools/resource-viewer')
    * features: [none]
  M serde_json 1.0.64 (direct third-party, crates.io)
    * added features: arbitrary_precision
    * (unchanged features: default, std)
```

**Comment by libra-action** [comment]
**LSR dependency change summary**:
```
target packages:
  A borrow-graph 0.0.1 (workspace, path 'language/borrow-graph')
    * features: [none]
  A bytecode-verifier 0.1.0 (workspace, path 'language/bytecode-verifier')
    * features: default
  A move-binary-format 0.1.0 (workspace, path 'language/move-binary-format')
    * features: default
  A move-vm-runtime 0.1.0 (workspace, path 'language/move-vm/runtime')
    * features: default
  A move-vm-types 0.1.0 (workspace, path 'language/move-vm/types')
    * features: default
  A fail 0.4.0 (direct third-party, crates.io)
    * features: [none]
  A petgraph 0.5.1 (direct third-party, crates.io)
    * features: default, graphmap, matrix_graph, stable_graph
  A sha3 0.9.1 (direct third-party, crates.io)
    * features: default, std
  A block-padding 0.2.1 (transitive third-party, crates.io)
    * features: [none]
  A fixedbitset 0.2.0 (transitive third-party, crates.io)
    * features: [none]
  A getrandom 0.1.16 (transitive third-party, crates.io) (other versions: 0.2.2)
    * features: std
  A hashbrown 0.9.1 (transitive third-party, crates.io)
    * features: raw
  A indexmap 1.6.2 (transitive third-party, crates.io)
    * features: [none]
  A keccak 0.1.0 (transitive third-party, crates.io)
    * features: [none]
  A rand 0.7.3 (transitive third-party, crates.io) (other versions: 0.8.3)
    * features: alloc, default, getrandom, getrandom_package, libc, std
  A rand_chacha 0.2.2 (transitive third-party, crates.io) (other versions: 0.3.0)
    * features: std
  A rand_core 0.5.1 (transitive third-party, crates.io) (other versions: 0.6.2)
    * features: alloc, getrandom, std
  M smallvec 1.6.1 (direct third-party, crates.io)
    * status changed from transitive third-party
    * (unchanged features: [none])
  M block-buffer 0.9.0 (transitive third-party, crates.io)
    * added features: block-padding
    * (unchanged features: [none])

host packages:
  A variant_count 1.1.0 (direct third-party, crates.io)
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A resource-viewer 0.1.0 (workspace, path 'language/tools/resource-viewer')
    * features: [none]
  M serde_json 1.0.64 (direct third-party, crates.io)
    * added features: arbitrary_precision
    * (unchanged features: default, std)
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A resource-viewer 0.1.0 (workspace, path 'language/tools/resource-viewer')
    * features: [none]
  M serde_json 1.0.64 (direct third-party, crates.io)
    * added features: arbitrary_precision
    * (unchanged features: default, std)
```

**Comment by libra-action** [comment]
**LSR dependency change summary**:
```
target packages:
  A borrow-graph 0.0.1 (workspace, path 'language/borrow-graph')
    * features: [none]
  A bytecode-verifier 0.1.0 (workspace, path 'language/bytecode-verifier')
    * features: default
  A move-binary-format 0.1.0 (workspace, path 'language/move-binary-format')
    * features: default
  A move-vm-runtime 0.1.0 (workspace, path 'language/move-vm/runtime')
    * features: default
  A move-vm-types 0.1.0 (workspace, path 'language/move-vm/types')
    * features: default
  A fail 0.4.0 (direct third-party, crates.io)
    * features: [none]
  A petgraph 0.5.1 (direct third-party, crates.io)
    * features: default, graphmap, matrix_graph, stable_graph
  A sha3 0.9.1 (direct third-party, crates.io)
    * features: default, std
  A block-padding 0.2.1 (transitive third-party, crates.io)
    * features: [none]
  A fixedbitset 0.2.0 (transitive third-party, crates.io)
    * features: [none]
  A getrandom 0.1.16 (transitive third-party, crates.io) (other versions: 0.2.2)
    * features: std
  A hashbrown 0.9.1 (transitive third-party, crates.io)
    * features: raw
  A indexmap 1.6.2 (transitive third-party, crates.io)
    * features: [none]
  A keccak 0.1.0 (transitive third-party, crates.io)
    * features: [none]
  A rand 0.7.3 (transitive third-party, crates.io) (other versions: 0.8.3)
    * features: alloc, default, getrandom, getrandom_package, libc, std
  A rand_chacha 0.2.2 (transitive third-party, crates.io) (other versions: 0.3.0)
    * features: std
  A rand_core 0.5.1 (transitive third-party, crates.io) (other versions: 0.6.2)
    * features: alloc, getrandom, std
  M smallvec 1.6.1 (direct third-party, crates.io)
    * status changed from transitive third-party
    * (unchanged features: [none])
  M block-buffer 0.9.0 (transitive third-party, crates.io)
    * added features: block-padding
    * (unchanged features: [none])

host packages:
  A variant_count 1.1.0 (direct third-party, crates.io)
    * features: [none]
```

File: language/tools/resource-viewer/src/lib.rs (hunk 1 above)
Comment (bmwill) [reply language/tools/resource-viewer/src/lib.rs]: Right, the client could just request the raw resource instead of this rendered one in that case. I guess I was over-thinking it :)

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A resource-viewer 0.1.0 (workspace, path 'language/tools/resource-viewer')
    * features: [none]
```

**Comment by libra-action** [comment]
**LSR dependency change summary**:
```
target packages:
  A borrow-graph 0.0.1 (workspace, path 'language/borrow-graph')
    * features: [none]
  A bytecode-verifier 0.1.0 (workspace, path 'language/bytecode-verifier')
    * features: default
  A move-binary-format 0.1.0 (workspace, path 'language/move-binary-format')
    * features: default
  A move-vm-runtime 0.1.0 (workspace, path 'language/move-vm/runtime')
    * features: default
  A move-vm-types 0.1.0 (workspace, path 'language/move-vm/types')
    * features: default
  A fail 0.4.0 (direct third-party, crates.io)
    * features: [none]
  A petgraph 0.5.1 (direct third-party, crates.io)
    * features: default, graphmap, matrix_graph, stable_graph
  A sha3 0.9.1 (direct third-party, crates.io)
    * features: default, std
  A block-padding 0.2.1 (transitive third-party, crates.io)
    * features: [none]
  A fixedbitset 0.2.0 (transitive third-party, crates.io)
    * features: [none]
  A getrandom 0.1.16 (transitive third-party, crates.io) (other versions: 0.2.2)
    * features: std
  A hashbrown 0.9.1 (transitive third-party, crates.io)
    * features: raw
  A indexmap 1.6.2 (transitive third-party, crates.io)
    * features: [none]
  A keccak 0.1.0 (transitive third-party, crates.io)
    * features: [none]
  A rand 0.7.3 (transitive third-party, crates.io) (other versions: 0.8.3)
    * features: alloc, default, getrandom, getrandom_package, libc, std
  A rand_chacha 0.2.2 (transitive third-party, crates.io) (other versions: 0.3.0)
    * features: std
  A rand_core 0.5.1 (transitive third-party, crates.io) (other versions: 0.6.2)
    * features: alloc, getrandom, std
  M smallvec 1.6.1 (direct third-party, crates.io)
    * status changed from transitive third-party
    * (unchanged features: [none])
  M block-buffer 0.9.0 (transitive third-party, crates.io)
    * added features: block-padding
    * (unchanged features: [none])

host packages:
  A variant_count 1.1.0 (direct third-party, crates.io)
    * features: [none]
```

**Comment by libra-action** [comment]
**LSR dependency change summary**:
```
target packages:
  A borrow-graph 0.0.1 (workspace, path 'language/borrow-graph')
    * features: [none]
  A bytecode-verifier 0.1.0 (workspace, path 'language/bytecode-verifier')
    * features: default
  A move-binary-format 0.1.0 (workspace, path 'language/move-binary-format')
    * features: default
  A move-vm-runtime 0.1.0 (workspace, path 'language/move-vm/runtime')
    * features: default
  A move-vm-types 0.1.0 (workspace, path 'language/move-vm/types')
    * features: default
  A fail 0.4.0 (direct third-party, crates.io)
    * features: [none]
  A petgraph 0.5.1 (direct third-party, crates.io)
    * features: default, graphmap, matrix_graph, stable_graph
  A sha3 0.9.1 (direct third-party, crates.io)
    * features: default, std
  A block-padding 0.2.1 (transitive third-party, crates.io)
    * features: [none]
  A fixedbitset 0.2.0 (transitive third-party, crates.io)
    * features: [none]
  A getrandom 0.1.16 (transitive third-party, crates.io) (other versions: 0.2.2)
    * features: std
  A hashbrown 0.9.1 (transitive third-party, crates.io)
    * features: raw
  A indexmap 1.6.2 (transitive third-party, crates.io)
    * features: [none]
  A keccak 0.1.0 (transitive third-party, crates.io)
    * features: [none]
  A rand 0.7.3 (transitive third-party, crates.io) (other versions: 0.8.3)
    * features: alloc, default, getrandom, getrandom_package, libc, std
  A rand_chacha 0.2.2 (transitive third-party, crates.io) (other versions: 0.3.0)
    * features: std
  A rand_core 0.5.1 (transitive third-party, crates.io) (other versions: 0.6.2)
    * features: alloc, getrandom, std
  M smallvec 1.6.1 (direct third-party, crates.io)
    * status changed from transitive third-party
    * (unchanged features: [none])
  M block-buffer 0.9.0 (transitive third-party, crates.io)
    * added features: block-padding
    * (unchanged features: [none])

host packages:
  A variant_count 1.1.0 (direct third-party, crates.io)
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A resource-viewer 0.1.0 (workspace, path 'language/tools/resource-viewer')
    * features: [none]
```

File: json-rpc/src/methods.rs [hunk 2]
```diff
@@ -388,6 +393,20 @@ impl<'a> Handler<'a> {
         data::get_network_status(self.service.role.as_str())
     }
 
+    /// Returns all resources in the account specified by `params`
+    async fn get_resources(
+        &self,
+        params: GetResourcesParams,
+    ) -> Result<BTreeMap<String, AnnotatedMoveStruct>, JsonRpcError> {
```
Comment (bmwill) [inline json-rpc/src/methods.rs:400]: I suppose this is intended to just return an empty map in the case where the account doesn't exist?

File: language/tools/resource-viewer/src/lib.rs [hunk 3]
```diff
@@ -226,6 +226,69 @@ fn pretty_print_ability_modifiers(f: &mut Formatter, abilities: AbilitySet) -> s
...
+            Vector(t, vals) => {
+                assert_ne!(t, &TypeTag::U8);
+                let mut vec = serializer.serialize_seq(Some(vals.len()))?;
+                for v in vals {
+                    vec.serialize_element(v)?;
+                }
+                vec.end()
+            }
+            Bytes(v) => {
+                // try to deserialize as utf8, fall back to hex with 0x prefix if we can't
```
Comment (bmwill) [inline language/tools/resource-viewer/src/lib.rs:274]: Are utf8 strings ever going to become first class citizens in move?

File: json-rpc/src/methods.rs (hunk 2 above)
Comment (sblackshear) [reply json-rpc/src/methods.rs:400]: Correct. We could return an error instead, but there's no concept of an account "existing" without containing some resources, so empty seems reasonable?

File: language/tools/resource-viewer/src/lib.rs (hunk 3 above)
Comment (sblackshear) [reply language/tools/resource-viewer/src/lib.rs:274]: Hacks like this make me think it's time to finally implement a string type :).

**Comment by libra-action** [comment]
**LSR dependency change summary**:
```
target packages:
  A borrow-graph 0.0.1 (workspace, path 'language/borrow-graph')
    * features: [none]
  A bytecode-verifier 0.1.0 (workspace, path 'language/bytecode-verifier')
    * features: default
  A move-binary-format 0.1.0 (workspace, path 'language/move-binary-format')
    * features: default
  A move-vm-runtime 0.1.0 (workspace, path 'language/move-vm/runtime')
    * features: default
  A move-vm-types 0.1.0 (workspace, path 'language/move-vm/types')
    * features: default
  A fail 0.4.0 (direct third-party, crates.io)
    * features: [none]
  A petgraph 0.5.1 (direct third-party, crates.io)
    * features: default, graphmap, matrix_graph, stable_graph
  A sha3 0.9.1 (direct third-party, crates.io)
    * features: default, std
  A block-padding 0.2.1 (transitive third-party, crates.io)
    * features: [none]
  A fixedbitset 0.2.0 (transitive third-party, crates.io)
    * features: [none]
  A getrandom 0.1.16 (transitive third-party, crates.io) (other versions: 0.2.2)
    * features: std
  A hashbrown 0.9.1 (transitive third-party, crates.io)
    * features: raw
  A indexmap 1.6.2 (transitive third-party, crates.io)
    * features: [none]
  A keccak 0.1.0 (transitive third-party, crates.io)
    * features: [none]
  A rand 0.7.3 (transitive third-party, crates.io) (other versions: 0.8.3)
    * features: alloc, default, getrandom, getrandom_package, libc, std
  A rand_chacha 0.2.2 (transitive third-party, crates.io) (other versions: 0.3.0)
    * features: std
  A rand_core 0.5.1 (transitive third-party, crates.io) (other versions: 0.6.2)
    * features: alloc, getrandom, std
  M smallvec 1.6.1 (direct third-party, crates.io)
    * status changed from transitive third-party
    * (unchanged features: [none])
  M block-buffer 0.9.0 (transitive third-party, crates.io)
    * added features: block-padding
    * (unchanged features: [none])

host packages:
  A variant_count 1.1.0 (direct third-party, crates.io)
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A resource-viewer 0.1.0 (workspace, path 'language/tools/resource-viewer')
    * features: [none]
```

File: language/tools/resource-viewer/src/lib.rs (hunk 3 above)
Comment (bmwill) [reply language/tools/resource-viewer/src/lib.rs:274]: There have been a number of times i've told folks to do string validation onchain so having some sort of string type would have made that easier/possible :)

File: json-rpc/integration-tests/src/lib.rs [hunk 4]
```diff
@@ -1730,6 +1730,65 @@ impl PublicUsageTest for GetEventByVersionWithProofTest {
...
+                "value": 4294967296_u64
+              },
+              "is_synthetic": true,
+              "scaling_factor": 1000000,
+              "fractional_part": 1000,
+              "currency_code": "XDX",
+              "can_mint": false,
+              "mint_events": {
+                "counter": 0,
+                "guid": "0x0a000000000000000000000000000000000000000a550c18"
```
Comment (xli) [inline json-rpc/integration-tests/src/lib.rs:1767]: although it's nice to see this format more consistent with other blockchain like Ethereum, but it seems we are creating a new format of hex-encoded string, as other APIs return hex-encoded string without "0x".

File: json-rpc/src/data.rs [hunk 5]
```diff
@@ -291,3 +293,25 @@ pub fn get_account_state_with_proof(
...
     )?)
 }
+
+/// Get all resources stored under `account_address` at `version`
+pub fn get_resources(
+    db: &dyn MoveDbReader,
+    ledger_version: u64,
+    account_address: AccountAddress,
+    version: u64,
+) -> Result<BTreeMap<String, AnnotatedMoveStruct>, JsonRpcError> {
```
Comment (xli) [inline json-rpc/src/data.rs:303]: I wonder why this function does not call to new function created for get_resource, instead of doing same thing by calling get_account_state_with_proof

File: json-rpc/src/methods.rs (hunk 2 above)
Comment (xli) [reply json-rpc/src/methods.rs:400]: we don't return error for get_account when the address is not an account address, so return empty is consistent with existing api

File: json-rpc/src/tests/utils.rs [hunk 6]
```diff
@@ -383,6 +387,52 @@ impl DbReader for MockDiemDB {
     }
 }
 
+impl MoveStorage for MockDiemDB {
```
Comment (xli) [inline json-rpc/src/tests/utils.rs:390]: seems sad to re-implement logic here.
Maybe a decorator is better?
Like:  MoveDbReader: MoveStorage { dbreader: DbReader }

File: language/tools/resource-viewer/src/lib.rs [hunk 7]
```diff
@@ -226,6 +226,69 @@ fn pretty_print_ability_modifiers(f: &mut Formatter, abilities: AbilitySet) -> s
     Ok(())
 }
```
Comment (xli) [inline language/tools/resource-viewer/src/lib.rs:228]: are we sure the test covers all serialization logic below?

File: language/tools/resource-viewer/src/lib.rs [hunk 8]
```diff
@@ -226,6 +226,69 @@ fn pretty_print_ability_modifiers(f: &mut Formatter, abilities: AbilitySet) -> s
...
 
+impl serde::Serialize for AnnotatedMoveStruct {
+    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        if serializer.is_human_readable() {
+            let mut s = serializer.serialize_map(Some(self.value.len()))?;
+            for (f, v) in &self.value {
+                s.serialize_entry(f, v)?
+            }
+            s.end()
+        } else {
```
Comment (xli) [inline language/tools/resource-viewer/src/lib.rs:237]: do we have something use the following part code? assuming get_resources used human readable part.

File: json-rpc/integration-tests/src/lib.rs (hunk 4 above)
Comment (xli) [reply json-rpc/integration-tests/src/lib.rs:1767]: I probably would like to start to use "0x" prefix for hex-encoded data in new API, but in json-rpc API, it maybe better to keep it consistent with other apis. Otherwise, the guid extracted here, probably can't be used in get_events api

File: json-rpc/integration-tests/src/lib.rs [hunk 9]
```diff
@@ -1730,6 +1730,65 @@ impl PublicUsageTest for GetEventByVersionWithProofTest {
...
+    fn run<'t>(&self, ctx: &mut PublicUsageContext<'t>) -> Result<()> {
+        let env = JsonRpcTestHelper::new(ctx.url().to_owned());
+        let address = format!("{:x}", diem_sdk::types::account_config::diem_root_address());
+        let response = env.send("get_resources", json!([address]));
+
+        let value = response.result.unwrap();
+        let resources = value.as_object().unwrap();
+        // check that the CurrencyInfo<XDX> resource looks ok as a sanity check
+        let currency_info = resources
+            .get("0x1::Diem::CurrencyInfo<0x1::XDX::XDX>")
```
Comment (xli) [inline json-rpc/integration-tests/src/lib.rs:1750]: what it looks like when there are multiple types? because the type_params in StructTag is an array.

File: json-rpc/integration-tests/src/lib.rs (hunk 4 above)
Comment (sblackshear) [reply json-rpc/integration-tests/src/lib.rs:1767]: That makes sense; I will change this. I wanted to have a prefix to distinguish hex strings from utf8 strings, but since we're not already doing this it doesn't make sense to start now. Perhaps worth thinking about in the REST API, though.

File: json-rpc/src/data.rs (hunk 5 above)
Comment (sblackshear) [reply json-rpc/src/data.rs:303]: Ah, it's because `get_resource` fetches the entire state blob + proof and then returns a single resource. I didn't want to fetch the state blob N times for an account with N resources, so I wrote slightly more code here instead of reusing `get_resource`.

File: json-rpc/src/tests/utils.rs (hunk 6 above)
Comment (sblackshear) [reply json-rpc/src/tests/utils.rs:390]: Agreed re: the sadness, I'll see if I can make the decorator work.

File: language/tools/resource-viewer/src/lib.rs (hunk 7 above)
Comment (sblackshear) [reply language/tools/resource-viewer/src/lib.rs:228]: It definitely does not; let me add some more cases to the E2E (or figure out if there's a more lightweight way to unit test this).

File: json-rpc/integration-tests/src/lib.rs (hunk 9 above)
Comment (sblackshear) [reply json-rpc/integration-tests/src/lib.rs:1750]: Same syntax as generics in Rust; it has a comma separator. E.g.,

`0x1::Module::Type<0x2::M::TypeArg1, 0x3::N::TypeArg2>`

File: language/tools/resource-viewer/src/lib.rs (hunk 8 above)
Comment (sblackshear) [reply language/tools/resource-viewer/src/lib.rs:237]: Good point; I think this is unused. Will eliminate.

File: json-rpc/src/tests/utils.rs (hunk 6 above)
Comment (sblackshear) [reply json-rpc/src/tests/utils.rs:390]: I tried, but couldn't do this without adding new public types in the storage interfaces that were only used in this test code. Stuck with the code duplication instead.

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A resource-viewer 0.1.0 (workspace, path 'language/tools/resource-viewer')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A resource-viewer 0.1.0 (workspace, path 'language/tools/resource-viewer')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A resource-viewer 0.1.0 (workspace, path 'language/tools/resource-viewer')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A resource-viewer 0.1.0 (workspace, path 'language/tools/resource-viewer')
    * features: [none]
```

**Comment by sblackshear** [comment]
I rebased this on top of https://github.com/diem/diem/commit/844c4a77980028e37edaeb8b52ca9ef9ff568a95, which makes things much nicer but not requiring the storage layer or json-rpc to depend on the VM. Think this should be good to go now.

File: json-rpc/Cargo.toml [hunk 10]
```diff
@@ -17,7 +17,7 @@ hex = "0.4.3"
 hyper = "0.14.4"
 once_cell = "1.7.2"
 rand = "0.8.3"
-serde_json = "1.0.64"
+serde_json = { version = "1.0.64" }
```
Comment (bmwill) [inline json-rpc/Cargo.toml:20]: Seems like an unnecessary change?

File: language/move-vm/runtime/src/data_cache.rs [hunk 11]
```diff
@@ -187,6 +187,8 @@ impl<'r, 'l, S: MoveResolver> DataStore for TransactionDataCache<'r, 'l, S> {
                 Ok(None) => GlobalValue::none(),
                 Err(err) => {
                     let msg = format!("Unexpected storage error: {:?}", err);
+                    // REVIEW: better way to get info out of a PartialVMError?
```
Comment (bmwill) [inline language/move-vm/runtime/src/data_cache.rs:190]: Is this intended to be resolved before this is landed or more as a TODO?

**Review by bmwill — COMMENTED** [review]
Looks great! just a single nit and a quick question and we can get this landed. Assuming @xli doesn't have any other comments.

File: language/move-vm/runtime/src/data_cache.rs (hunk 11 above)
Comment (sblackshear) [reply language/move-vm/runtime/src/data_cache.rs:190]: This is actually a copy/paste error--will fix.

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A resource-viewer 0.1.0 (workspace, path 'language/tools/resource-viewer')
    * features: [none]
```

**Review by bmwill — APPROVED** [review]
